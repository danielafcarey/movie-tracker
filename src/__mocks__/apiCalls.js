import { mockMovies } from '../mockData';

export const fetchMovies = jest.fn().mockImplementation( () => {
  return Promise.resolve( mockMovies );
});
