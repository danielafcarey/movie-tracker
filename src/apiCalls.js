import { apiKey } from './apiKey';

const fetchMovies = async () => {
  const url = `https://api.themoviedb.org/3/movie/now_playing?api_key=${apiKey}`;
  try {
    const response = await fetch(url);
    if (response.status === 200) {
      const data = await response.json();
      return data;
    } else {
      throw Error(`Status failure: ${response.status}`);
    }
  } catch (error) {
    throw error;
  }
}; 

const fetchUsers = async () => {
  const url = 'http://localhost:3000/api/users';
  try {
    const response = await fetch(url);
    if (response.status === 200) {
      const data = await response.json();
      return data.data;
    } else {
      throw Error(response.status);
    }
  } catch (error) {
    throw Error(error);
  }

};

const fetchFavorites = async (id) => {
  const url = `http://localhost:3000/api/users/${id}/favorites`;
  try {
    const response = await fetch(url);
    if (response.status === 200) {
      const data = await response.json();
      return data.data;
    } else {
      throw Error(response.status);
    }
  } catch (error) {
    throw Error(error);
  }
}

const postFavorite = async (userId, movie) => {
  const url = `http://localhost:3000/api/users/favorites/new`;
  const favorite = Object.assign({}, movie, { user_id: userId, overview: '' });
  const optionsObject = {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(favorite)
  };

  try {
    const response = await fetch(url, optionsObject);  
    if (response.status !== 200) {
      throw Error(response.status);
    }
  } catch (error) {
    throw Error(error);
  }
}

const deleteFavorite = async (userId, movieId) => {
  const url = `http://localhost:3000/api/users/${userId}/favorites/${movieId}`; 
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error(response.status)
    }
  } catch(error) {
    throw Error(error);
  }
}

export {
  fetchMovies,
  fetchUsers,
  fetchFavorites,
  postFavorite,
  deleteFavorite
};

