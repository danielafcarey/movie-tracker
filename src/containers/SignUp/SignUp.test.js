import React from 'react';
import { shallow } from 'enzyme';
import { 
  SignUp,
  mapDispatchToProps
} from './SignUp';
import * as apiCalls from '../../apiCalls';

describe('SignUp', () => {
  let wrapper;
  let mockUpdateCurrentUser;

  beforeEach(() => {
    mockUpdateCurrentUser = jest.fn()
    wrapper = shallow(<SignUp updateCurrentUser={mockUpdateCurrentUser}/>);
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has a default state of name, email, password, verification, emailError, and passwordError', () => {
    expect(wrapper.state('name')).toEqual('');
    expect(wrapper.state('email')).toEqual('');
    expect(wrapper.state('password')).toEqual('');
    expect(wrapper.state('verification')).toEqual('');
    expect(wrapper.state('emailError')).toEqual('');
    expect(wrapper.state('passwordError')).toEqual('');
    expect(wrapper.state('authenticated')).toEqual(false);
  });

  it('changes the name in state on change in the name input', () => {
    const mockEvent = { target: { 
      value: 'garbage', 
      name: 'name'
    } };

    wrapper.instance().handleChange(mockEvent);

    expect(wrapper.state('name')).toEqual('garbage');
  });

  it('changes the email in state on change in the email input', () => {
    const mockEvent = { target: { 
      value: 'garbage', 
      name: 'email'
    } };

    wrapper.instance().handleChange(mockEvent);

    expect(wrapper.state('email')).toEqual('garbage');
  });

  it('changes the password in state on change in the password input', () => {
    const mockEvent = { target: { 
      value: 'garbage', 
      name: 'password'
    } };

    wrapper.instance().handleChange(mockEvent);

    expect(wrapper.state('password')).toEqual('garbage');
  });

  it('changes the verification in state on change in the verification input', () => {
    const mockEvent = { target: { 
      value: 'garbage', 
      name: 'verification'
    } };

    wrapper.instance().handleChange(mockEvent);

    expect(wrapper.state('verification')).toEqual('garbage');
  });

  describe('verifyPassword', () => {

    it('returns true and sets state if the password and verification match', () => {
      const mockState = {
        name: 'Nincompoop',
        email: 'nincompooping@gmail.com',
        password: 'ilovebabiesandgarbage',
        verification: 'ilovebabiesandgarbage',
        passwordError: 'Passwords must match'
      };

      wrapper.setState({ ...mockState });
      const result = wrapper.instance().verifyPassword();

      expect(result).toEqual(true);
      expect(wrapper.state('passwordError')).toEqual('');
    });

    it('returns false and sets state if the password does not match the verification', () => {
      const mockState = {
        name: 'Nincompoop',
        email: 'nincompooping@gmail.com',
        password: 'ilovebabiesandgarbage',
        verification: 'ihatebabiesbutilovegarbage'
      };

      wrapper.setState({ ...mockState });
      const result = wrapper.instance().verifyPassword();

      expect(result).toEqual(false);
      expect(wrapper.state('passwordError')).toEqual('Passwords must match');

    });

  });

  describe('verifyEmail', () => {

    it('calls fetchUsers', async () => {
      apiCalls.fetchUsers = jest.fn().mockImplementation(() => {
        return [
          {
            id: 1,
            email: 'garbageman@gmail.com',
            password: 'ismellbad',
            name: 'jerry'
          }
        ]
      })
      await wrapper.instance().verifyEmail();
      expect(apiCalls.fetchUsers).toHaveBeenCalled()
    });

    it('returns true if the user\'s email is not yet taken', async () => {
      wrapper.setState({ 
        email: 'foolsgold@gmail.com',
        emailError: 'Email has already been used'
      });

      const result =  await wrapper.instance().verifyEmail();

      expect(result).toEqual(true);
      expect(wrapper.state('emailError')).toEqual('');
    });

    it('returns false and sets state if the user\'s email is taken', async () => {
      wrapper.setState({ email: 'garbageman@gmail.com' });

      const result =  await wrapper.instance().verifyEmail();

      expect(result).toEqual(false);
      expect(wrapper.state('emailError')).toEqual('Email has already been used')
    });

  });


  describe('postUser', () => {

    it('calls fetch with the correct arguments', () => {
      window.fetch = jest.fn().mockImplementation(() => {
        return Promise.resolve({
          status: 200,
          json: () => Promise.resolve(1)
        })
      });
      const mockState = {
        name: 'nincompoop',
        email: 'nincompooping@gmail.com',
        password: 'ilovebabiesandgarbage'
      };
      const expectedUrl = 'http://localhost:3000/api/users/new';
      const expectedOptionsObject = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(mockState) 
      };
      
      wrapper.setState({ ...mockState });
      wrapper.instance().postUser();

      expect(window.fetch).toHaveBeenCalledWith(expectedUrl, expectedOptionsObject);
    });

  });

  describe('handleSubmit', () => {
    let mockEvent;

    beforeEach(() => {
      mockEvent = { preventDefault : () => {}}
    })

      it('calls verifyPassword', () => {
        const wrapperInst = wrapper.instance();
        wrapperInst.verifyPassword = jest.fn();
        wrapperInst.verifyEmail = jest.fn();
        wrapperInst.postUser = jest.fn();

        wrapperInst.handleSubmit(mockEvent);

        expect(wrapperInst.verifyPassword).toHaveBeenCalled();
      });
      
      it('calls updateCurrentUser with the correct arguments if password and email have been verified', async () => {
        const wrapperInst = wrapper.instance();
  
        wrapperInst.verifyPassword = jest.fn().mockImplementation(() => true);
        wrapperInst.verifyEmail = jest.fn().mockImplementation(() => true);
        wrapperInst.postUser = jest.fn().mockImplementation(() => 1);

        await wrapperInst.handleSubmit(mockEvent);

        expect(wrapperInst.props.updateCurrentUser).toHaveBeenCalledWith(1);
      });
      
      it('calls postUser with the correct arguments if password has been verified', async () => {
        const wrapperInst = wrapper.instance();
        wrapperInst.verifyPassword = jest.fn().mockImplementation(() => true);
        wrapperInst.verifyEmail = jest.fn().mockImplementation(() => true);
        wrapperInst.postUser = jest.fn();
        
        await wrapperInst.handleSubmit(mockEvent);

        expect(wrapperInst.postUser).toHaveBeenCalled();
      });

    it('sets state of authenticated to true', async () => {
      const wrapperInst = wrapper.instance();
      wrapperInst.verifyPassword = jest.fn().mockImplementation(() => true);
      wrapperInst.verifyEmail = jest.fn().mockImplementation(() => true);
      wrapperInst.postUser = jest.fn();

      await wrapperInst.handleSubmit(mockEvent);

      expect(wrapper.state('authenticated')).toEqual(true);
    }) 
       
    });

  describe('mapDispatchToProps', () => {
    
    it('returns an object with an updateCurrentUser function', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);

      expect(typeof result.updateCurrentUser).toEqual('function');
    });
  });

});






















