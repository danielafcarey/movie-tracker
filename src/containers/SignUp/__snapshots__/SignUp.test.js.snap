// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUp
    updateCurrentUser={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <input
                name="name"
                onChange={[Function]}
                placeholder="Name*"
                required={true}
                type="text"
                value=""
        />
        <input
                name="email"
                onChange={[Function]}
                placeholder="Email*"
                required={true}
                type="text"
                value=""
        />
        <p />
        <input
                name="password"
                onChange={[Function]}
                placeholder="Password*"
                required={true}
                type="password"
                value=""
        />
        <input
                name="verification"
                onChange={[Function]}
                placeholder="Retype password*"
                required={true}
                type="password"
                value=""
        />
        <p />
        <button>
                Sign Up
        </button>
</form>,
      "className": "signup",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            name="name"
            onChange={[Function]}
            placeholder="Name*"
            required={true}
            type="text"
            value=""
/>,
          <input
            name="email"
            onChange={[Function]}
            placeholder="Email*"
            required={true}
            type="text"
            value=""
/>,
          <p />,
          <input
            name="password"
            onChange={[Function]}
            placeholder="Password*"
            required={true}
            type="password"
            value=""
/>,
          <input
            name="verification"
            onChange={[Function]}
            placeholder="Retype password*"
            required={true}
            type="password"
            value=""
/>,
          <p />,
          <button>
            Sign Up
</button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "name",
            "onChange": [Function],
            "placeholder": "Name*",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "email",
            "onChange": [Function],
            "placeholder": "Email*",
            "required": true,
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "password",
            "onChange": [Function],
            "placeholder": "Password*",
            "required": true,
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "verification",
            "onChange": [Function],
            "placeholder": "Retype password*",
            "required": true,
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign Up",
          },
          "ref": null,
          "rendered": "Sign Up",
          "type": "button",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <input
                    name="name"
                    onChange={[Function]}
                    placeholder="Name*"
                    required={true}
                    type="text"
                    value=""
          />
          <input
                    name="email"
                    onChange={[Function]}
                    placeholder="Email*"
                    required={true}
                    type="text"
                    value=""
          />
          <p />
          <input
                    name="password"
                    onChange={[Function]}
                    placeholder="Password*"
                    required={true}
                    type="password"
                    value=""
          />
          <input
                    name="verification"
                    onChange={[Function]}
                    placeholder="Retype password*"
                    required={true}
                    type="password"
                    value=""
          />
          <p />
          <button>
                    Sign Up
          </button>
</form>,
        "className": "signup",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              name="name"
              onChange={[Function]}
              placeholder="Name*"
              required={true}
              type="text"
              value=""
/>,
            <input
              name="email"
              onChange={[Function]}
              placeholder="Email*"
              required={true}
              type="text"
              value=""
/>,
            <p />,
            <input
              name="password"
              onChange={[Function]}
              placeholder="Password*"
              required={true}
              type="password"
              value=""
/>,
            <input
              name="verification"
              onChange={[Function]}
              placeholder="Retype password*"
              required={true}
              type="password"
              value=""
/>,
            <p />,
            <button>
              Sign Up
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "name",
              "onChange": [Function],
              "placeholder": "Name*",
              "required": true,
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "email",
              "onChange": [Function],
              "placeholder": "Email*",
              "required": true,
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "password",
              "onChange": [Function],
              "placeholder": "Password*",
              "required": true,
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "verification",
              "onChange": [Function],
              "placeholder": "Retype password*",
              "required": true,
              "type": "password",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sign Up",
            },
            "ref": null,
            "rendered": "Sign Up",
            "type": "button",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
